src/ejabberd_c2s.erl

```diff
@@ -2980,16 +3002,34 @@ stream_mgmt_request() ->
 flush_stream_mgmt_buffer(#state{stream_mgmt = StreamMgmt})
   when StreamMgmt =:= false; StreamMgmt =:= disabled ->
     false;
-flush_stream_mgmt_buffer(#state{stream_mgmt_buffer = Buffer}) ->
-    re_route_packets(Buffer).
-
-re_route_packets(Buffer) ->
-    [begin
-         {From, To, _El} = mongoose_acc:packet(Acc),
-         ejabberd_router:route(From, To, Acc)
-     end || Acc <- lists:reverse(Buffer)],
+flush_stream_mgmt_buffer(#state{stream_mgmt_buffer = Buffer} = State) ->
+    re_route_packets(Buffer, State).
+
+re_route_packets(Buffer, StateData) ->
+    [reroute(Acc, StateData) || Acc <- lists:reverse(Buffer)],
     ok.
 
+reroute(Acc, StateData) ->
+    {From, To, _El} = mongoose_acc:packet(Acc),
+    reroute(From, To, Acc, StateData).
+
+reroute(From, To, Acc, #state{sid = SID}) ->
+    Acc2 = patch_acc_for_reroute(Acc, SID),
+    ejabberd_router:route(From, To, Acc2).
+
+patch_acc_for_reroute(Acc, SID) ->
+    case mongoose_acc:stanza_name(Acc) of
+        <<"message">> ->
```

Comment:
We also need to not reroute MAM results messages (i.e. reroute, but ignore them, if SID is wrong), not just IQs and presences.
MAM results could confuse a reconnected client too.

But this would require another test.